utils/communication_report.py

```diff
@@ -0,0 +1,133 @@
+    new_url = f"{base_url}\?_getpages={getpages}\&_getpagesoffset={getpagesoffset}\&_count={count}\&_bundletype={bundletype}"
+    return new_url
+
+def get_patient_isacc_id(patient_reference, patient_cache):
+    # Check if we already have the ISACC ID for this patient
+    if patient_reference in patient_cache:
+        return patient_cache[patient_reference]
+
+    patient_url = f"{base_url}/{patient_reference}"
+    patient_data = get_fhir_resource(patient_url)
+
+    # Initialize ISACC ID as empty string
+    isacc_id = ""
+
+    # Search for ISACC ID in the patient's identifiers
+    for identifier in patient_data.get('identifier', []):
+        if identifier.get('system') == "http://isacc.app/user-id":
+            isacc_id = identifier.get('value', "")
+            break
+
+    # Cache the ISACC ID for future use
+    patient_cache[patient_reference] = isacc_id
+    return isacc_id
+
+def extract_type(communication):
+    for category in communication.get('category', []):
+        for coding in category.get('coding', []):
+            if coding.get('system') == "https://isacc.app/CodeSystem/communication-type":
+                return coding.get('code')
+    return None
+
+def main():
+    url = f"{base_url}/Communication?_sort=sent"
+    patient_cache = {}
+
+    with open('CommunicationReport.csv', 'w', newline='') as file:
+        writer = csv.writer(file)
+        writer.writerow(["ID", "Datetime", "Recipient FHIR ID", "Recipient ISACC ID", "Type", "Content", "Sender", "Note"])
+
+        while url:
+            data = get_fhir_resource(url)
+            for entry in data.get('entry', []):
+                communication = entry['resource']
+                communication_id = communication.get('id')
+                datetime = communication.get('sent')
+                type = extract_type(communication)
+                if type in ['isacc-comment', 'isacc-non-sms-message']:
+                    recipient_reference = communication.get('subject', {}).get('reference')
+                else:
+                    recipient_reference = communication.get('recipient', [{}])[0].get('reference')
+                if recipient_reference == None:
+                    recipient_reference = communication.get('sender', {}).get('reference')
+                isacc_id = get_patient_isacc_id(recipient_reference, patient_cache)
+                content = communication.get('payload', [{}])[0].get('contentString')
+                if content:
+                    content = content.replace('\n', '\\n')
+                sender = communication.get('sender', {}).get('reference')
+                if sender == None and type == 'isacc-auto-sent-message':
+                    sender = 'system'
+                note = communication.get('note', [{}])[0].get('text')
+                if note:
+                    note = note.replace('\n', '\\n')
+
+                writer.writerow([communication_id, datetime, recipient_reference, isacc_id, type, content, sender, note])
```

Comment:
might make sense to move the Communication processing/extraction to a separate function (and return a row/list), so it's easier to keep track of the fields of interest